tap2audio.1
.TH tap2audio 1 "17th May, 2003"
.SH NAME
tap2audio \- convert a .TAP file to a Commodore 64 tape
.SH SYNOPSIS
tap2audio -h|-v
.P
tap2audio [-i] [-f <freq>] [-v <vol>] [-w <triangle|square|sine>] input_TAP_file_name [output_WAV_file_name]
.SH DESCRIPTION
.I tap2audio
reads the input .TAP or .DMP file and converts it to a Commodore 64 tape. The output
format can be a sound (this is the default) or a .WAV file (if the name of
a WAV file is specified as second argument).
\fIlibtapdecoder.so\fR is required.
Conversion to sound card requires \fIlibportaudio.so.2\fR.
Conversion to WAV file requires \fIlibaudiofile.so.0\fR.

.SH OPTIONS
.TP
.B -f <freq>
Output waveform will have frequency <freq> Hz.
.TP
.B -h
Show a help message and exit successfully.
.TP
.B -i
Generate an inverted waveform. Some soundcards invert the signal
when playing back, i.e. they turn positive waveform values into negative ones
and vice versa. If a normal waveform is played back with such a soundcard, a
C64 will not be able to recognize it. If an inverted waveform is played back
with such a soundcard, the soundcard will invert it again, turning it into a
normal waveform, so a C64 will recognize it. If a waveform generated without the
.B -i
option is not loaded by the C64, try enabling this option. This option is ignored
if the input is a TAP version 2 (C16 TAP file with semiwave recording).
.TP
.B -v <vol>
Set volume of output waveform to <vol> (1-255).
.TP
.B -V
Show version and copyright info and exit successfully
.TP
.B -w <triangle|square|sine>
Use triangular/square/sinusoidal waveform (default: square)
.SH AUTHOR
Fabrizio Gennari (fabrizio.ge@tiscalinet.it)
.SH REFERENCES
.I http://www.computerbrains.com/tapformat.html
.IP
Per Hakan Sundell's description of .TAP file format
.PP
.I http://www.portaudio.com
.IP
PortAudio library home page. \fItap2audio\fR needs it for playing back C64 tapes
to sound card
.PP
.I http://www.68k.org/~michael/audiofile
.IP
Audiofile library home page. \fItap2audio\fR needs it for generating C64 tapes
as WAV files
.SH SEE ALSO
.B wav2prg
(1),
.B prg2wav
(1),
.B audio2tap
(1)


tap2audio.c
/* Audiotap: a program for playing TAP files or converting them to WAV files,
 * and to convert C64 tapes to TAP files
 *
 * Copyright (c) Fabrizio Gennari, 2003
 *
 * The program is distributed under the GNU General Public License.
 * See file LICENSE.TXT for details.
 *
 * tap2audio.c : main tap->audio program for the command-line version
 * 
 * This file belongs to the tap->audio part
 * This file is part of the command-line version of Audiotap
 */

#include <stdlib.h>
#include <stdio.h>
#include <getopt.h>
#include <signal.h>
#include <string.h>
#include <sys/types.h>

#include "audiotap.h"
#include "tap2audio_core.h"
#include "audiotap_loop.h"

static void sig_int(int signum){
  audiotap_interrupt();
}

void help(){
  printf("Usage: tap2audio -h|-V\n");
  printf("       tap2audio [-f <freq>] [-v vol] [-i] <input TAP file> [output WAV file]\n");
  printf("Options:\n");
  printf("\t-h: show this help message and exit successfully\n");
  printf("\t-V: show version and copyright info and exit successfully\n");
  printf("\t-i: use inverted waveforms (ignored is input is TAP version 2)\n");
  printf("\t-f <freq>: use output frequency <freq> Hz (default:44100)\n");
  printf("\t-v: volume of the output sound (0-255, default:254)\n");
  printf("\t-w <square|triangle|sine>: set waveform (default:square)\n");
  printf("If no output WAV file is specified, output is sound card\n");
}

void version(){
  printf("tap2audio (part of Audiotap) version 1.5\n");
  printf("(C) by Fabrizio Gennari, 2003-2008\n");
  printf("This program is distributed under the GNU General Public License\n");
  printf("Read the file LICENSE.TXT for details\n");
  printf("This product includes software developed by the NetBSD\n");
  printf("Foundation, Inc. and its contributors\n");
}
   
int main(int argc, char** argv){
  struct tapdec_params params = {254, TAP_TRIGGER_ON_RISING_EDGE, TAPDEC_SQUARE};
  int freq = 44100;
  struct option cmdline[]={
    {"volume"            ,1,NULL,'v'},
    {"frequency"         ,1,NULL,'f'},
    {"inverted-waveform" ,0,NULL,'i'},
    {"waveform"          ,1,NULL,'w'},
    {"help"              ,0,NULL,'h'},
    {"version"           ,0,NULL,'V'},
    {NULL                ,0,NULL,0}
  };
  int option;
  struct audiotap_init_status status;

  status = audiotap_initialize();
  if ((status.audiofile_init_status != LIBRARY_OK &&
      status.portaudio_init_status != LIBRARY_OK)
   || status.tapdecoder_init_status != LIBRARY_OK){
    printf("Failed to initialize audiotap library: both audiofile and pablio failed to load");
    exit(1);
  }

  while( (option=getopt_long(argc,argv,"v:f:ihVw:",cmdline,NULL)) != -1){
    switch(option){
    case 'v':
      params.volume=atoi(optarg);
      if (params.volume < 1){
        printf("Volume too low\n");
        exit(1);
      };
      break;
    case 'f':
      freq=atoi(optarg);
      break;
    case 'i':
      params.inverted=1;
      break;
    case 'h':
      help();
      exit(0);
    case 'V':
      version();
      exit(0);
    case 'w':
      if(!strcmp(optarg,"square"))
        params.waveform = TAPDEC_SQUARE;
      else if(!strcmp(optarg,"sine"))
        params.waveform = TAPDEC_SINE;
      else if(!strcmp(optarg,"triangle"))
        params.waveform = TAPDEC_TRIANGLE;
      else{
        printf("Wrong argument to option -c\n");
        exit(1);
      }
      break;
    default:
      help();
      exit(1);
    }
  }

  argc-=optind;
  argv+=optind;

  if(argc < 1){
    printf("You must specify an input TAP file!\n");
    exit(1);
  }

  if(argc > 2){
    printf("Too many arguments\n");
    exit(1);
  }

  if (argc == 1){
    if (status.portaudio_init_status != LIBRARY_OK){
      printf("Cannot read from sound card: pablio library missing or invalid\n");
      exit(1);
    }
  }
  else{
    if (status.audiofile_init_status != LIBRARY_OK){
      printf("Cannot read from file: audiofile library missing or invalid\n");
      exit(1);
    }
  }

  signal(SIGINT, sig_int);

  tap2audio(argv[0], argv[1], &params, freq);
  exit(0);
}

